src/stdlib_selection.fypp

```diff
@@ -0,0 +1,275 @@
+
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("select", 1, arraytype, arraykind, intkind)
+      module procedure ${name}$
+    #:endfor
+  #:endfor
+end interface
+
+interface arg_select
+    !! version: experimental
+    !! ([Specification](..//page/specs/stdlib_selection.html#arg_select-find-the-index-of-the-k-th-smallest-value-in-an-input-array))
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("arg_select", 1, arraytype, arraykind, intkind)
+      module procedure ${name}$
+    #:endfor
+  #:endfor
+end interface
+
+contains
+
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("select", 1, arraytype, arraykind, intkind)
+      subroutine ${name}$(a, k, kth_smallest, left, right)
+          !! select - select the k-th smallest entry in a(:).
+          !!
+          !! Partly derived from the "Coretran" implementation of 
+          !! quickSelect by Leon Foks, https://github.com/leonfoks/coretran
+          !!
+          ${arraytype}$, intent(inout) :: a(:)
+              !! Array in which we seek the k-th smallest entry.
+              !! On output it will be partially sorted such that
+              !! `all(a(1:(k-1)) <= a(k)) .and. all(a(k) <= a((k+1):size(a)))`.
+          ${inttype}$, intent(in) :: k
+              !! We want the k-th smallest entry. E.G. `k=1` leads to
+              !! `kth_smallest=min(a)`, and `k=size(a)` leads to
+              !! `kth_smallest=max(a)`
+          ${arraytype}$, intent(out) :: kth_smallest
+              !! On output contains the k-th smallest value of `a(:)`
+          ${inttype}$, intent(in), optional :: left, right
+              !! If we know that:
+              !!    the k-th smallest entry of `a` is in `a(left:right)`
+              !! and also that:
+              !!    `maxval(a(1:(left-1))) <= minval(a(left:right))`
+              !! and:
+              !!    `maxval(a(left:right))) <= minval(a((right+1):size(a)))`
+              !! then one or both bounds can be specified to narrow the search.
+              !! The constraints are available if we have previously called the
+              !! subroutine with different `k` (because of how `a(:)` becomes
+              !! partially sorted, see documentation for `a(:)`).
+
+          ${inttype}$ :: l, r, mid, iPivot
+          integer, parameter :: ip = ${intkind}$
+
+          l = 1_ip
+          if(present(left)) l = left
+          r = size(a, kind=ip)
+          if(present(right)) r = right
+
+          if(k < 1_ip .or. k > size(a, kind=ip) .or. l > r .or. l < 1_ip .or. &
+              r > size(a, kind=ip)) then
+              error stop "select must have 1 <= k <= size(a), and 1 <= left <= right <= size(a)";
+          end if
+
+          do while(.true.)
```

Comment:
```suggestion
          do
```

In case of simple `do` loops like this, I like to name them.

Reply:
Tried this but decided that, because the loop is so short, the name didn't help me. But I have removed the `while(.true.)`

Reply:
There was a discussion recently on exiting loops: https://fortran-lang.discourse.group/t/ending-a-do-loop/2329/6?u=ivanpribec

The suggestion from John Collins who works on the static analysis tool [fpt](http://simconglobal.com/fpt_summary.html) was to never leave an exit un-tagged.

Reply:
Addressed in a commit just now -- this loop is labelled, as well as the ones containing `exit`.